test/Conversion/VectorToAIEVec/test-vector-fma.mlir

```diff
@@ -0,0 +1,37 @@
+// RUN: aie-opt %s -convert-vector-to-aievec="aie-target=aie2" -split-input-file | FileCheck %s
+
+// CHECK-LABEL: test_fma_bf16
+// CHECK-SAME: %[[V0:[a-zA-Z0-9]+]]: vector<16xbf16>,
+// CHECK-SAME: %[[V1:.*]]: vector<16xbf16>,
+// CHECK-SAME: %[[V2:.*]]: vector<16xbf16>)
+func.func @test_fma_bf16(%v0: vector<16xbf16>,
+                         %v1: vector<16xbf16>,
+                         %v2: vector<16xbf16>) -> vector<16xbf16> {
+  // CHECK: %[[ACC:.*]] = aievec.ups %[[V2]] {shift = 0 : i8}
+  // CHECK-SAME:              : vector<16xbf16>, vector<16xf32>
+  // CHECK: %[[FMA:.*]] = aievec.mac_elem %[[V0]], %[[V1]], %[[ACC]]
+  // CHECK-SAME:              : vector<16xbf16>, vector<16xbf16>, vector<16xf32>
+  // CHECK: %[[RES:.*]] = aievec.srs %[[FMA]], %{{[a-zA-Z0-9]+}}
+  // CHECK-SAME:              : vector<16xf32>, i32, vector<16xbf16>
+  // CHECK: return %[[RES]] : vector<16xbf16>
+  %0 = vector.fma %v0, %v1, %v2 : vector<16xbf16>
+  return %0 : vector<16xbf16>
+}
+
+// -----
+
+// CHECK-LABEL: test_fma_f32
+// CHECK-SAME: %[[V0:[a-zA-Z0-9]+]]: vector<16xbf16>,
+// CHECK-SAME: %[[V1:.*]]: vector<16xbf16>,
+// CHECK-SAME: %[[V2:.*]]: vector<16xf32>)
+func.func @test_fma_f32(%v0: vector<16xbf16>,
```

Comment:
The 16 lanes referred to in the pattern, does that mean only fma with vectors of 16 elements is supported?

Reply:
For the matmuls, I use the table https://github.com/Xilinx/mlir-aie/blob/ea17ecba6006c6a33ce38844d0237be8f38bce47/include/aie/Dialect/AIEVec/IR/AIEVecOps.td#L870 to help tiling in IREE. Is there something similar for fma? Or is it as simple as  "fma must be 256 bits"?

Reply:
Fma must be 256bit (2x16 bfloat16 elements into a single vector of 16 f32 elements). It should be possible to combine a sequence of two linked 256bit fma into a single 2x256bit fma, but I'd have to experiment a bit to make sure I'm not breaking anything.